### PR TITLE
block: Add query_device_size() and fix size handling for block devices

### PR DESCRIPTION
**Pull request by weltling**
### Background

During the block I/O refactoring in #7892 (and accompanying #7885), a latent bug was revealed. When a block device (`/dev/loopN`, device-mapper snapshot, etc.) is passed as a disk image, both `logical_size()` and `physical_size()` return wrong values.

On Linux, `st_size` in `metadata().len()` is always 0 for block devices, and `st_blocks` is also 0. The old code used `seek(SeekFrom::End(0))` which happened to work, but the refactored `DiskSize` trait requires immutable `&self`, making seek unavailable. Switching to `metadata().len()` exposed the bug - guests see a disk with 0 capacity.

Three CI tests fail on block device passthrough paths: `test_virtio_block_discard_loop_device`, `test_virtio_block_discard_dm_snapshot`, and `test_virtio_block_direct_io_block_device_alignment_4k`.

### Solution

This series introduces `query_device_size(file: &File) -> io::Result<(u64, u64)>` which returns `(logical_size, physical_size)` using a single `fstat` call:

- Regular files: logical = `st_size`, physical = `st_blocks * 512` (actual host allocation, correct for sparse files)
- Block devices: both = `BLKGETSIZE64` ioctl (correct, works with immutable &File)
- Other fd types: returns `InvalidInput` error

All raw backends (`raw_async`, `raw_sync`, `raw_async_aio`) and the qcow `RawFile` are updated to use it for both `logical_size()` and `physical_size()`.

Additionally, the advisory byte range lock in `virtio-device`s was using `physical_size()` for its lock range. With the corrected size reporting, neither `physical_size()` nor `logical_size()` alone is sufficient in all cases: sparse files can have `logical > physical`, while small dense files can have `physical > logical` due to filesystem block rounding. The lock now uses `max(logical, physical)` so the locked range safely covers both cases.

### CODE REVIEW

File: block/src/lib.rs [hunk 1]
```diff
@@ -1193,6 +1194,33 @@ ioctl_io_nr!(BLKSSZGET, 0x12, 104);
...
+        // st_blocks is always in 512-byte units on Linux
+        Ok((m.len(), m.st_blocks() * 512))
+    } else if m.file_type().is_block_device() {
+        let mut size: u64 = 0;
+        // SAFETY: BLKGETSIZE64 reads the device size into a u64 pointer.
+        let ret = unsafe { libc::ioctl(file.as_raw_fd(), BLKGETSIZE64() as _, &mut size) };
+        if ret != 0 {
+            return Err(io::Error::last_os_error());
+        }
+        Ok((size, size))
```
Comment (russell-islam) [inline block/src/lib.rs:1216]: For block device, logical and physical size are same? Why?? Just for my understanding.

File: block/src/lib.rs [hunk 2]
```diff
@@ -1193,6 +1194,33 @@ ioctl_io_nr!(BLKSSZGET, 0x12, 104);
...
 ioctl_io_nr!(BLKIOMIN, 0x12, 120);
 ioctl_io_nr!(BLKIOOPT, 0x12, 121);
+ioctl_ior_nr!(BLKGETSIZE64, 0x12, 114, u64);
+
+/// Returns `(logical_size, physical_size)` in bytes for regular files and block devices.
+///
+/// For regular files, logical size is `st_size` and physical size is
+/// `st_blocks * 512` (actual host allocation). For block devices both
+/// values equal the `BLKGETSIZE64` result.
+pub fn query_device_size(file: &File) -> io::Result<(u64, u64)> {
```
Comment (russell-islam) [inline block/src/lib.rs:1204]: Why not BlockResult?

File: block/src/lib.rs (hunk 1 above)
Comment (weltling) [reply block/src/lib.rs:1216]: Thanks for checking. Block devices have no sparse allocation. Every byte of the device exists physically. That's why also `st_blocks` was delivering 0.

Thanks

File: block/src/lib.rs (hunk 2 above)
Comment (weltling) [reply block/src/lib.rs:1204]: This is a low level utility operating on `&File` directly, so `io::Result` is the natural fit here. `BlockResult` is meant for the boundary with higher level consumers. Using `io::Result` avoids having to rewrite callers sitting deeper in the stack. They wrap it in their own error type as needed, like `DiskFileError::Size`, `RawFileError` and so on.

Thanks

File: block/src/lib.rs [hunk 3]
```diff
@@ -1193,6 +1194,33 @@ ioctl_io_nr!(BLKSSZGET, 0x12, 104);
...
+    let m = file.metadata()?;
+    if m.is_file() {
+        // st_blocks is always in 512-byte units on Linux
+        Ok((m.len(), m.st_blocks() * 512))
```
Comment (phip1611) [inline block/src/lib.rs:1208]: regarding `512`. After a first glance, I couldn't finde a constant in the libc crate for that. So we can keep it as is

File: block/src/lib.rs (hunk 2 above)
Comment (phip1611) [inline block/src/lib.rs:1204]: I wonder if you can do a very simple unit test for that using `/tmp/foo` and `/dev/zero` 🤔

File: block/src/lib.rs [hunk 4]
```diff
@@ -1193,6 +1194,33 @@ ioctl_io_nr!(BLKSSZGET, 0x12, 104);
...
+        // SAFETY: BLKGETSIZE64 reads the device size into a u64 pointer.
+        let ret = unsafe { libc::ioctl(file.as_raw_fd(), BLKGETSIZE64() as _, &mut size) };
+        if ret != 0 {
+            return Err(io::Error::last_os_error());
+        }
+        Ok((size, size))
+    } else {
+        Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "disk image must be a regular file or block device",
```
Comment (phip1611) [inline block/src/lib.rs]: `... is: {:?}", m.file_type()`

File: block/src/lib.rs (hunk 3 above)
Comment (russell-islam) [inline block/src/lib.rs:1208]: Even for that file, cant we use m.blksize()

File: virtio-devices/src/block.rs [hunk 5]
```diff
@@ -888,7 +888,7 @@ impl Block {
             LockGranularityChoice::Full => LockGranularity::WholeFile,
             LockGranularityChoice::ByteRange => {
                 // Byte-range lock covering [0, size)
-                self.disk_image.physical_size().map_or_else(
+                self.disk_image.logical_size().map_or_else(
```
Comment (phip1611) [inline virtio-devices/src/block.rs]: I am unsure if one way or the other is correct for all disk types/images 🤔 

Also, what does Linux say if the byte range is larger than the physical size?

File: block/src/lib.rs (hunk 3 above)
Comment (weltling) [reply block/src/lib.rs:1208]: No, `blksize()` returns `st_blksize` from the `stat` struct. That is the preferred I/O block size for the filesystem typically 4096, not the device capacity. It is the size that the kernel would wish to be used for I/O, it doesn't tell the actual device parameters.

Thanks

Comment (weltling) [reply block/src/lib.rs:1208]: Yep, this is `st_blocks` from `stat` which is always at 512 byte granularity.

Thanks

File: block/src/lib.rs (hunk 2 above)
Comment (weltling) [reply block/src/lib.rs:1204]: Gotcha. I first thought it's actually covered through existing tests, but this topic is far more complicated than it seemed to be. I've added some more sophisticated unit tests just for this new function, including the `/dev/zero` one. The block device is then covered by the integration tests in any case.

Thanks

File: block/src/lib.rs (hunk 4 above)
Comment (weltling) [reply block/src/lib.rs]: Yeah, that's nicer. Fixed, thanks.

File: virtio-devices/src/block.rs (hunk 5 above)
Comment (weltling) [reply virtio-devices/src/block.rs]: Thanks for the second thought. Actually yes, `physical > logical` is possible, too, due to filesystem block rounding. I rewrote this to take `max(logical, physical)`.

Linux advisory locks are purely inode metadata in the kernel lock table - no I/O is performed and the kernel does not validate the range against the actual file size. From `fcntl_locking(2)` subpage:

> Bytes past the end of the file may be locked, but not bytes before the start of the file.

Thanks

Comment (phip1611) [reply virtio-devices/src/block.rs]: Sounds like a pragmatic solution to the problem!

**Comment by rbradford** [comment]
@phip1611 Your newly added test seems to flake out on the MQ on this one :-( - https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/23631631446/job/68832189770

**Comment by phip1611** [comment]
> @phip1611 Your newly added test seems to flake out on the MQ on this one :-( - https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/23631631446/job/68832189770

Thanks! I'll look into it
